## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [23]:
# View the column names in the data
df.columns
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [22]:
# Use the describe function to gather some basic statistics
statistics = df.describe()  
print(statistics)   

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [24]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
top_three_categories = category_counts.head(3)
print(top_three_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [25]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df['category'].value_counts().idxmax()
filtered_df = df[df['category'] == most_common_category]
subcategory_counts = filtered_df['subcategory'].value_counts()
most_common_subcategory = subcategory_counts.idxmax()
print(f"The category with the most entries is: {most_common_category}")
print(f"The subcategory with the most entries within {most_common_category} is: {most_common_subcategory}")

The category with the most entries is: consumables
The subcategory with the most entries within consumables is: bathroom supplies


In [63]:
# Which five clients had the most entries in the data?
top_five_clients = client_counts.head(5)
top_five_client_ids = df[df['full_name'].isin(top_five_clients.index)][['full_name', 'client_id']].drop_duplicates()
display(top_five_client_ids)



full_name  client_id
5      Jessica Reyes      33615
137      Bryan Myers      46820
216   Angela Everett      66037
416   Kendra Garrett      24741
470  Alexandra Young      38378

In [64]:
# Store the client ids of those top 5 clients in a list.
df['full_name'] = df['first'] + ' ' + df['last']
client_counts = df['full_name'].value_counts()
top_five_clients = client_counts.head(5).index
top_five_client_ids = df[df['full_name'].isin(top_five_clients)]['client_id'].unique().tolist()
print(top_five_client_ids)

[33615, 46820, 66037, 24741, 38378]


In [31]:
# How many total units (the qty column) did the client with the most entries order order?
df['full_name'] = df['first'] + ' ' + df['last']
most_common_client = df['full_name'].value_counts().idxmax()
client_df = df[df['full_name'] == most_common_client]
total_units_ordered = client_df['qty'].sum()
print(f"The total units ordered by the client with the most entries ({most_common_client}) is: {total_units_ordered}")




The total units ordered by the client with the most entries (Jessica Reyes) is: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [65]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df = pd.read_csv('resources/client_dataset.csv')
df['subtotal'] = df['unit_price'] * df['qty']
display(df.head())

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [66]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
display(df.head())


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [42]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df = pd.read_csv('resources/client_dataset.csv')
df['full_name'] = df['first'] + ' ' + df['last']
df['subtotal'] = df['unit_price'] * df['qty']
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
display(df)

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number           full_name   subtotal  \
0             7.50   105            1      Donald Harding  115164.00   
1             1.49    21            0       Tiffany Myers     523.95   
2             1.68    39            6      Shannon Watson     527.28   
3             1.23    29            3        Nathan Baker    1056.18   
4            46.43    20            1  Christina Schwartz    3902.00   
...            ...   ...          ...                 ...        ...   
54634         2.25    33            8        Lauren Reese    2743.29   
54635        11.70    47            0       Derrick Moore    9709.73   
54636         4.16   475            6    Monica Gutierrez   31188.50   
54637        18.04   112            0       Wanda Solomon     165.76   
54638         2.07  1031            5         Dwayne Bell    3103.31   

      shipping_price    total_price  
0               75.0  125898.607500  
1               14.9     588.693625  
2               16.8     594.407400  
3               12.3    1167.314400  
4              464.3    4770.182750  
...              ...            ...  
54634           22.5    3021.625575  
54635          117.0   10735.70

In [48]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df = pd.read_csv('resources/client_dataset.csv')
df['full_name'] = df['first'] + ' ' + df['last']
df['subtotal'] = df['unit_price'] * df['qty']
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
display(df)

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number           full_name   subtotal  shipping_price  \
0                1      Donald Harding  115164.00            75.0   
1                0       Tiffany Myers     523.95            14.9   
2                6      Shannon Watson     527.28            16.8   
3                3        Nathan Baker    1056.18            12.3   
4                1  Christina Schwartz    3902.00           464.3   
...            ...                 ...        ...             ...   
54634            8        Lauren Reese    2743.29            22.5   
54635            0       Derrick Moore    9709.73           117.0   
54636            6    Monica Gutierrez   31188.50            41.6   
54637            0       Wanda Solomon     165.76           180.4   
54638            5         Dwayne Bell    3103.31            20.7   

         total_price  line_cost  
0      125898.607500   80159.55  
1         588.693625     331.79  
2         594.407400     323.34  
3        1167.314400     732.95  
4        4770.182750    2627.70  
...              ...        ...  
54634    3021.625575    1725.30  
54635   10735.702525    8363.62  
54636   34118.884250   27263.85  
54637     378.179800     317.04  
54638    3412.980925 

In [51]:
# Create a column for the profit of each line using line cost and line price
df = pd.read_csv('resources/client_dataset.csv')
df['full_name'] = df['first'] + ' ' + df['last']
df['subtotal'] = df['unit_price'] * df['qty']
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df['line_profit'] = df['total_price'] - df['line_cost']
display(df)

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

                full_name   subtotal  shipping_price    total_price line_cost  \
0          Donald Harding  115164.00            75.0  125898.607500  80159.55   
1           Tiffany Myers     523.95            14.9     588.693625    331.79   
2          Shannon Watson     527.28            16.8     594.407400    323.34   
3            Nathan Baker    1056.18            12.3    1167.314400    732.95   
4      Christina Schwartz    3902.00           464.3    4770.182750   2627.70   
...                   ...        ...             ...            ...       ...   
54634        Lauren Reese    2743.29            22.5    3021.625575   1725.30   
54635       Derrick Moore    9709.73           117.0   10735.702525   8363.62   
54636    Monica Gutierrez   31188.50            41.6   34118.884250  27263.85   
54637       Wanda Solomon     165.76           180.4     378.179800    317.04   
54638         Dwayne Bell    3103.31            20.7    3412.980925   1680.61   

        line_profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637  

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [69]:
# Check your work using the totals above
# Load your dataset into a DataFrame
df = pd.read_csv('resources/client_dataset.csv')
df['full_name'] = df['first'] + ' ' + df['last']

# Create a subtotal column
df['subtotal'] = df['unit_price'] * df['qty']

# Create a shipping price column
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)

# Calculate the total price including sales tax
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925

# Calculate the cost of each line
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Calculate the profit of each line
df['line_profit'] = df['total_price'] - df['line_cost']

# Define the order IDs and their expected total prices
orders = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Verify the total prices for each order ID
for order_id, expected_total_price in orders.items():
    order_df = df[df['order_id'] == order_id]
    total_price_order = order_df['total_price'].sum()
    print(f"Total price for Order ID {order_id}: {total_price_order}")
    print(f"Does the calculated total price match the expected total price? {total_price_order == expected_total_price}")


Total price for Order ID 2742071: 139844.51202499997
Does the calculated total price match the expected total price? False
Total price for Order ID 2173913: 119838.35704999999
Does the calculated total price match the expected total price? False
Total price for Order ID 6128929: 752748.7454750001
Does the calculated total price match the expected total price? False


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [52]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
